graphql_compiler/schema_transformation/merge_schemas.py

```diff
@@ -161,14 +152,100 @@ def _get_basic_schema_ast(query_type):
+          types/interfaces/enums/scalars, or conflicts between the definition of directives
+          with the same name
+    """
+    # Check input schema identifier is a string of alphanumeric characters and underscores
+    check_schema_identifier_is_valid(current_schema_id)
+    # Check input schema satisfies various structural requirements
+    check_ast_schema_is_valid(current_ast)
+
+    current_schema = build_ast_schema(current_ast)
+    current_query_type = get_query_type_name(current_schema)
+
+    # Merge current_ast into merged_schema_ast.
+    # Concatenate new scalars, new directives, and type definitions other than the query
+    # type to definitions list.
+    # Raise errors for conflicting scalars, directives, or types.
+    new_definitions = current_ast.definitions  # List[Node]
+    new_query_type_fields = None  # List[FieldDefinition]
+
+    for new_definition in new_definitions:
+        if isinstance(new_definition, ast_types.SchemaDefinition):
+            continue
+        elif (
+            isinstance(new_definition, ast_types.ObjectTypeDefinition) and
+            new_definition.name.value == current_query_type
+        ):  # query type definition
+            new_query_type_fields = new_definition.fields  # List[FieldDefinition]
+        elif isinstance(new_definition, ast_types.DirectiveDefinition):
+            _process_directive_definition(
+                new_definition, directives, merged_schema_ast
+            )
+        elif isinstance(new_definition, ast_types.ScalarTypeDefinition):
+            _process_scalar_definition(
+                new_definition, scalars, type_name_to_schema_id, merged_schema_ast
+            )
+        elif isinstance(new_definition, (
+            ast_types.EnumTypeDefinition,
+            ast_types.InterfaceTypeDefinition,
+            ast_types.ObjectTypeDefinition,
+            ast_types.UnionTypeDefinition,
+        )):
+            _process_generic_type_definition(
+                new_definition, current_schema_id, scalars, type_name_to_schema_id,
+                merged_schema_ast
+            )
+        else:  # All definition types should've been covered
+            raise AssertionError(
+                u'Missed definition type: "{}"'.format(type(new_definition).__name__)
+            )
+
+    # Concatenate all query type fields.
+    # Since query_type was taken from the schema built from the input AST, the query type
+    # should never be not found.
+    if new_query_type_fields is None:
+        raise AssertionError(u'Query type "{}" field definitions unexpected not '
+                             u'found.'.format(current_query_type))
+    # Note that as field names and type names have been confirmed to match up, and types
+    # were merged without name conflicts, query type fields can also be safely merged.
+    query_type_index = 1  # Query type is the second entry in the list of definitions
```

Comment:
This is guaranteed to be the case because _get_basic_schema_ast builds a nearly blank schema whose second entry is the query type. The merged schema AST starts off being that basic schema, and all future additions just append more type definitions to the end. There is a doc string in _get_basic_schema_ast that says the second entry is the query type, but I'll add another comment here explaining the reason since it's easy to miss.